src/Middleware/HttpLogging/src/W3CLoggingMiddleware.cs

```diff
@@ -0,0 +1,264 @@
+            {
+                throw new ArgumentNullException(nameof(options));
+            }
+
+            if (w3cLogger == null)
+            {
+                throw new ArgumentNullException(nameof(w3cLogger));
+            }
+
+            _options = options;
+            _w3cLogger = w3cLogger;
+        }
+
+        /// <summary>
+        /// Invokes the <see cref="HttpLoggingMiddleware" />.
+        /// </summary>
+        /// <param name="context"></param>
+        /// <returns></returns>HttpResponseLog.cs
+        public async Task Invoke(HttpContext context)
+        {
+            var options = _options.CurrentValue;
+
+            var elements = new string[_fieldsLength];
+
+            var now = DateTime.Now;
+            if (options.LoggingFields.HasFlag(W3CLoggingFields.Date))
+            {
+                AddToList(elements, W3CLoggingFields.Date, now.ToString("yyyy-MM-dd", CultureInfo.InvariantCulture));
+            }
+
+            if (options.LoggingFields.HasFlag(W3CLoggingFields.Time))
+            {
+                AddToList(elements, W3CLoggingFields.Time, now.ToString("HH:mm:ss", CultureInfo.InvariantCulture));
+            }
+
+            if ((W3CLoggingFields.ConnectionInfoFields & options.LoggingFields) != W3CLoggingFields.None)
+            {
+                var connectionInfo = context.Connection;
+
+                if (options.LoggingFields.HasFlag(W3CLoggingFields.ClientIpAddress))
+                {
+                    AddToList(elements, W3CLoggingFields.ClientIpAddress, connectionInfo.RemoteIpAddress is null ? "" : connectionInfo.RemoteIpAddress.ToString());
+                }
+
+                if (options.LoggingFields.HasFlag(W3CLoggingFields.ServerIpAddress))
+                {
+                    AddToList(elements, W3CLoggingFields.ServerIpAddress, connectionInfo.LocalIpAddress is null ? "" : connectionInfo.LocalIpAddress.ToString());
+                }
+
+                if (options.LoggingFields.HasFlag(W3CLoggingFields.ServerPort))
+                {
+                    AddToList(elements, W3CLoggingFields.ServerPort, connectionInfo.LocalPort.ToString(CultureInfo.InvariantCulture));
+                }
+            }
+
+            if ((W3CLoggingFields.Request & options.LoggingFields) != W3CLoggingFields.None)
+            {
+                var request = context.Request;
+
+                if (options.LoggingFields.HasFlag(W3CLoggingFields.ProtocolVersion))
+                {
+                    AddToList(elements, W3CLoggingFields.ProtocolVersion, request.Protocol);
+                }
+
+                if (options.LoggingFields.HasFlag(W3CLoggingFields.Method))
+                {
+                    AddToList(elements, W3CLoggingFields.Method, request.Method);
+                }
+
+                if (options.LoggingFields.HasFlag(W3CLoggingFields.UriStem))
+                {
+                    AddToList(elements, W3CLoggingFields.UriStem, request.Path.Value);
```

Comment:
Path probably needs to be the escaped format originally requested by the client?
```suggestion
                    AddToList(elements, W3CLoggingFields.UriStem, request.Path.ToUriComponent());
```
ToString also works, but this is clearer about intent.